### PR TITLE
[Dist Debugality] Log key DDP metrics to stderr under debug mode.

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52966 [DDP Logging] Log comm. hook in ddp logging
* #52965 [DDP Logging] Log nccl_async_error_handling
* **#52957 [Dist Debugality] Log key DDP metrics to stderr under debug mode.**
* #52887 [Dist Debugability] Refactor and fix DDP model check during init

This diff:
1. Under TORCH_DISTRIBUTED_DEBUG=INFO or DETAIL, logs DDP information during init time (all stats in ddp_logging_data_)
2. Under TORCH_DISTRIBUTED_DEBUG=DETAIL, logs runtime stats when they are collected (first 10 iterations and then once every 100 iterations). Avoiding logging every iteration to not spam logs.
3. Dedupes the logic to set the above debug level as per request.

Verified by inspecting logs:

```
I0226 19:12:47.109243 2818475 logger.cpp:140] [Rank 1]: DDP Initialized with:
world_size: 2 module_name: Linear device_ids: 1 output_device: 1 backend_name: nccl parameter_dtype: float total
_parameter_size_in_bytes: 40 num_parameter_tensors: 2 bucket_sizes: 40 CUDA_VISIBLE_DEVICES: N/Abroadcast_buffer
s: 1 bucket_cap_mb: 25 find_unused_parameters: 0 gradient_as_bucket_view: 0
 Backend Info: nccl_socket_ifname: N/A nccl_blocking_wait: N/A nccl_debug: WARN nccl_nthreads: N/A nccl_ib_timeo
ut: N/A
I0226 19:12:47.109252 2818473 logger.cpp:140] [Rank 0]: DDP Initialized with:
world_size: 2 module_name: Linear device_ids: 0 output_device: 0 backend_name: nccl parameter_dtype: float total
_parameter_size_in_bytes: 40 num_parameter_tensors: 2 bucket_sizes: 40 CUDA_VISIBLE_DEVICES: N/Abroadcast_buffer
s: 1 bucket_cap_mb: 25 find_unused_parameters: 0 gradient_as_bucket_view: 0
 Backend Info: nccl_socket_ifname: N/A nccl_blocking_wait: N/A nccl_debug: WARN nccl_nthreads: N/A nccl_ib_timeo
ut: N/A
```

```
I0226 19:12:48.117936 2818473 logger.cpp:286] [Rank 0 / 2] Training Linear unused_parameter_size=0
 Avg forward compute time: 568944
 Avg backward compute time: 885504
Avg backward comm. time: 692496
 Avg backward comm/comp overlap time: 113536
I0226 19:12:48.118517 2818475 logger.cpp:286] [Rank 1 / 2] Training Linear unused_parameter_size=0
 Avg forward compute time: 565584
 Avg backward compute time: 876992
Avg backward comm. time: 201872
 Avg backward comm/comp overlap time: 128624
```

Differential Revision: [D26708184](https://our.internmc.facebook.com/intern/diff/D26708184/)

### CODE REVIEW

File: torch/lib/c10d/logger.cpp [hunk 1]
```diff
@@ -19,8 +20,45 @@ std::string parse_env(const char* env_var_name) {
   return res;
 }
 
+std::once_flag kDistDebugLoggingLevelSetFlag;
+
+const char * parseDistDebugLevel() {
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/lib/c10d/logger.cpp:25]: `get_debug_level()` is python only so we need a way to get it from C++ as well. Possibly we can dedupe and have the python layer call into these cpp functions.

File: torch/lib/c10d/logger.cpp [hunk 2]
```diff
@@ -19,8 +20,45 @@ std::string parse_env(const char* env_var_name) {
...
+  static std::string debugLevel = parse_env(kDistDebugEnvVar);
+  std::call_once(kDistDebugLoggingLevelSetFlag, []() {
+    LOG(INFO) << "TORCH_DISTRIBUTED_DEBUG level parsed as " << debugLevel;
+  });
+
+  if (!debugLevel.compare(("N/A"))) {
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/lib/c10d/logger.cpp:31]: `.compare()` returns 0 if they are equal, so !debuglevel.compare("n/a") means that they are equal. Maybe compare directly to 0 to make this clearer.

File: torch/lib/c10d/reducer.cpp [hunk 3]
```diff
@@ -16,6 +16,12 @@
 #include <torch/csrc/utils/memory.h>
 
 namespace c10d {
+
+const char* kDistDebugEnvVar = "TORCH_DISTRIBUTED_DEBUG";
+const char* kDistDebugDetailLogLevel = "DETAIL";
+const char* kDistDebugInfoLogLevel = "INFO";
+const char* kDistDebugOffLogLevel = "OFF";
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/lib/c10d/reducer.cpp:23]: Similar to above it's only available in python.

**Comment by zhaojuanmao** [comment]
@rohan-varma  could we move _get_debug_mode() to c10d/utils.hpp/cpp in C++, exposed it to python. So that all the applications (like DDP) can use the same helper function to read debug mode info in both python and C++?

**Comment by rohan-varma** [comment]
@zhaojuanmao Moved it to C++ and binded it to python, please take another look, thanks!

File: torch/lib/c10d/Utils.hpp [hunk 4]
```diff
@@ -31,6 +31,39 @@ typedef SSIZE_T ssize_t;
...
+    {kDistDebugOffLogLevel, DistributedDebugLevel::OFF},
+    {kDistDebugInfoLogLevel, DistributedDebugLevel::INFO},
+    {kDistDebugDetailLogLevel, DistributedDebugLevel::DETAIL},\
+  };
+
+  auto it = mapping.find(levelStr);
+  TORCH_CHECK(
+    it != mapping.end(),
+    "Invalid string value for distributed debug mode: ", levelStr
+  );
```
Comment (zhaojuanmao) [inline torch/lib/c10d/Utils.hpp:59]: nit: feel this check can be merged with the check in parseDistDebugLevel(), also this function can be merged with parseDistDebugLevel() as well

**Review by zhaojuanmao — APPROVED** [review]
please see one minor comment for merging check and function, otherwise it looks good!

**Pull request closed without merging**